src/json.c

```diff
@@ -187,66 +214,116 @@ int JSON_StoreVectorInDocField(FieldSpec *fs, RedisJSON arr, struct DocumentFiel
     return REDISMODULE_ERR;
   }
 
-  // The right function will put a value of the right type in the address given, or return REDISMODULE_ERR
-  switch (type) {
-    default:
-    case VecSimType_FLOAT32:
-      getElement = (getJSONElementFunc)JSON_getFloat32;
-      break;
-    // Uncomment when support for more types is added
-    // case VecSimType_FLOAT64:
-    //   getElement = (getJSONElementFunc)JSON_getFloat64;
-    //   break;
-    // case VecSimType_INT32:
-    //   getElement = (getJSONElementFunc)JSON_getInt32;
-    //   break;
-    // case VecSimType_INT64:
-    //   getElement = (getJSONElementFunc)japi->getInt;
-    //   break;
-  }
+  getElement = VecSimGetJSONCallback(type);
 
   if (!(df->strval = rm_malloc(fs->vectorOpts.expBlobSize))) {
     return REDISMODULE_ERR;
   }
   df->strlen = fs->vectorOpts.expBlobSize;
 
-  unsigned char step = VecSimType_sizeof(type);
-  char *offset = df->strval;
   // At this point array length matches blob length
-  for (int i = 0; i < arrLen; ++i) {
-    RedisJSON json = japi->getAt(arr, i);
-    if (getElement(json, offset) != REDISMODULE_OK) {
-      rm_free(df->strval);
-      return REDISMODULE_ERR;
-    }
-    offset += step;
+  if (JSON_StoreVectorAt(arr, arrLen, getElement, df->strval, VecSimType_sizeof(type)) != REDISMODULE_OK) {
+    rm_free(df->strval);
+    return REDISMODULE_ERR;
   }
   df->unionType = FLD_VAR_T_CSTR;
   return REDISMODULE_OK;
 }
 
-typedef enum {
-  ITERABLE_ITER = 0,
-  ITERABLE_ARRAY = 1
-} JSONIterableType;
-
-// An adapter for iterator operations, such as `next`, over an underlying container/collection or iterator
-typedef struct {
-  JSONIterableType type;
-  union {
-    JSONResultsIterator iter;
-    struct {
-      RedisJSON arr;
-      size_t index;
-    } array;
-  };
-} JSONIterable;
+int JSON_StoreMultiVectorInDocField(FieldSpec *fs, JSONIterable *itr, size_t len, struct DocumentField *df) {
+  VecSimType type;
+  size_t dim;
+  bool multi;
+  getJSONElementFunc getElement;
+  RedisJSON element;
+
+  switch (fs->vectorOpts.vecSimParams.algo) {
+    case VecSimAlgo_HNSWLIB:
+      type = fs->vectorOpts.vecSimParams.hnswParams.type;
+      dim = fs->vectorOpts.vecSimParams.hnswParams.dim;
+      multi = fs->vectorOpts.vecSimParams.hnswParams.multi;
+      break;
+    case VecSimAlgo_BF:
+      type = fs->vectorOpts.vecSimParams.bfParams.type;
+      dim = fs->vectorOpts.vecSimParams.bfParams.dim;
+      multi = fs->vectorOpts.vecSimParams.bfParams.multi;
+      break;
+    default: goto fail;
+  }
+
+  if (!multi)
+    goto fail;
+
+  getElement = VecSimGetJSONCallback(type);
+  unsigned char step = VecSimType_sizeof(type);
+
+  if (!(df->blobArr = rm_malloc(fs->vectorOpts.expBlobSize * len))) {
```

Comment:
is it ok that we will allocate more than we need?
```
[[v1 values], null, [v2 values]]
```
see the example above

Reply:
Yes, we store non-null values continuously. I'll add a test and a comment about that.